### PR TITLE
Build - refactor for bundle factoring and swappable runtime

### PR DESCRIPTION
**Pull request by kumavis**
this refactor unblocks incremental lavamoat protections for the webapp
this is **does not add lavamoat** yet

some additional benefits:
- automatically factors bundles into entry (phishing/ui/background) and "common" module groups
- module groups are then reduced to subgroups based on a size limit, and then compose into bundles 
- bundles call into a swappable runtime api that allows the same module definition to be used in lavamoat or as in here a normal commonjs runtime. eg background could use lavamoat protections while ui does not, even as both consume the same "common" bundles

additional `watchify` and `patch-package` note:
in order to remove an incompatible watchify feature (binary output statistics), we used `patch-package`. it was determined to be the easiest and least hacky way to disable the non-optional feature.

### CODE REVIEW

**Comment by kumavis** [comment]
stuck on some sourcemap bug

File: development/build/scripts.js [hunk 1]
```diff
@@ -225,12 +182,119 @@ function createScriptTasks({ browserPlatforms, livereload }) {
...
+}) {
+  return async function () {
+    // create bundler setup and apply defaults
+    const buildConfiguration = createBuildConfiguration();
+    const { bundlerOpts, events } = buildConfiguration;
+
+    // devMode options
+    const reloadOnChange = Boolean(devMode);
+    // const minify = !devMode;
+    const minify = false;
```
Comment (ryanml) [inline development/build/scripts.js:199]: Should this be tied to `devMode`?

File: development/stream-flat-map.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  async function* flatMapGenerator(source) {
+    for await (const entry of source) {
+      const subStream = entryToStream(entry);
+      yield* subStream;
+    }
+  }
+}
+
+// this stupid utility turns an async iterator factory into a duplex stream
```
Comment (ryanml) [inline development/stream-flat-map.js:25]: lmao

File: development/build/scripts.js (hunk 1 above)
Comment (kumavis) [reply development/build/scripts.js:199]: yes, nice catch

File: development/build/scripts.js [hunk 3]
```diff
@@ -69,72 +64,48 @@ function createScriptTasks({ browserPlatforms, livereload }) {
...
-        label,
-        destFilepath: `${label}.js`,
-        modulesToExpose: externalDependenciesMap[key],
-        devMode: false,
+  function createTasksForBuildJsExtension({ taskPrefix, devMode, testing }) {
+    const standardEntryPoints = ['background', 'ui', 'phishing-detect'];
+
+    // let extraEntries;
+    // if (devMode && label === 'ui') {
+    //   extraEntries = ['./development/require-react-devtools.js'];
```
Comment (kumavis) [inline development/build/scripts.js:78]: ah, need to address this

**Comment by kumavis** [comment]
hmm seems enabling minification broke sourcemaps

**Comment by darkwing** [comment]
I tried running `yarn setup && yarn start` and I'm getting:

```
scripts:core:dev:standardEntryPoints: error Command failed with exit code 1.
scripts:core:dev:standardEntryPoints: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
MetaMask build: Encountered an error while running task "dev".
Error: MetaMask build: runInChildProcess for task "scripts:core:dev:standardEntryPoints" encountered an error 1
    at ChildProcess.<anonymous> (/Users/davidwalsh/Projects/metamask-extension/development/build/task.js:97:13)
    at Object.onceWrapper (events.js:422:26)
    at ChildProcess.emit (events.js:315:20)
    at ChildProcess.EventEmitter.emit (domain.js:467:12)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:277:12)
error Command failed with exit code 1.
```

**Comment by kumavis** [comment]
```
TypeError#1: Invalid non-string/buffer chunk
  at validChunk (/home/gudahtt/projects/metamask-extension/node_modules/readable-stream/lib/_stream_writable.js:305:10)
  at DestroyableTransform.Writable.write (/home/gudahtt/projects/metamask-extension/node_modules/readable-stream/lib/_stream_writable.js:333:62)
  at PassThrough.ondata (/home/gudahtt/projects/metamask-extension/node_modules/readable-stream/lib/_stream_readable.js:620:20)
  at PassThrough.emit (events.js:315:20)
  at PassThrough.EventEmitter.emit (domain.js:486:12)
  at addChunk (/home/gudahtt/projects/metamask-extension/node_modules/readable-stream/lib/_stream_readable.js:292:12)
  at readableAddChunk (/home/gudahtt/projects/metamask-extension/node_modules/readable-stream/lib/_stream_readable.js:279:11)
  at PassThrough.Readable.push (/home/gudahtt/projects/metamask-extension/node_modules/readable-stream/lib/_stream_readable.js:246:10)
  at PassThrough.Transform.push (/home/gudahtt/projects/metamask-extension/node_modules/readable-stream/lib/_stream_transform.js:149:32)
  at PassThrough.afterTransform (/home/gudahtt/projects/metamask-extension/node_modules/readable-stream/lib/_stream_transform.js:92:10)
  ```

File: development/build/scripts.js [hunk 4]
```diff
@@ -225,12 +180,118 @@ function createScriptTasks({ browserPlatforms, livereload }) {
...
+      devMode,
+      envVars,
+      reloadOnChange,
+      minify,
+    });
+
+    // set bundle entries
+    bundlerOpts.entries = [...entryFiles];
+
+    // setup bify-vinyl-gator plugin
```
Comment (kumavis) [inline development/build/scripts.js:209]: rename plugin in comment

**Comment by kumavis** [comment]
hypothesis is `watchify` conflicting with the changes

**Comment by ryanml** [comment]
Hey @darkwing - can you go ahead and have another look at this? the error you mention should be resolved now

**Comment by ryanml** [comment]
Also note, the `watchify` patch is not directly related to SES

File: development/build/scripts.js (hunk 4 above)
Comment (kumavis) [reply development/build/scripts.js:209]:
```suggestion
    // setup bundle factoring with bify-module-groups plugin 
```

**Review by kumavis — COMMENTED** [review]
im happy with it now

File: development/stream-flat-map.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  const readableStream = new ThroughStream({ objectMode: true });
+  const duplex = duplexify(writableStream, readableStream);
+  const asyncIter = factoryFn(writableStream);
+  // drain iterator into readable stream
+  process.nextTick(async () => {
+    try {
+      for await (const item of asyncIter) {
+        readableStream.write(item);
+      }
+      console.log('ending');
```
Comment (brad-decker) [inline development/stream-flat-map.js:37]: Is this needed for future debugging? If not 
```suggestion
```

**Comment by kumavis** [comment]
- browserify `dedupe` option?
- metamask ci bot links (eg bundle viz)

**Comment by kumavis** [comment]
desire for stats on total shipped code size

**Review by brad-decker — DISMISSED** [review]
The changes we discussed are good to have but not required for this to land imo. Even the broken links in the build output bot notification are not an issue due to the all artifacts link working.

**Review by darkwing — DISMISSED** [review]

**Comment by kumavis** [comment]
is this the right pack https://github.com/MetaMask/metamask-extension/blob/26f767377b20d9de6f98dabbea2d861daf1465ed/development/build/scripts.js#L237

**Comment by kumavis** [comment]
verify `groupByFactor`

**Comment by ryanml** [comment]
Bundle links have been fixed, this could be formatted a bit cleaner but that can be resolved in a follow up

**Review by darkwing — APPROVED** [review]

**Review by brad-decker — APPROVED** [review]

**Comment by kumavis** [comment]
ok i noticed the `bify-module-groups` 1.0.x versions include breaking changes, so im locking it in at `1.0.0`

**Comment by kumavis** [comment]
ok im ready

**Pull request merged**